vllm/model_executor/models/jais.py

```diff
@@ -0,0 +1,416 @@
+            kv_cache=kv_cache,
+            input_metadata=input_metadata,
+        )
+        # residual connection
+        hidden_states = attn_output + residual
+
+        residual = hidden_states
+        hidden_states = self.ln_2(hidden_states)
+        feed_forward_hidden_states = self.mlp(hidden_states)
+        # residual connection
+        hidden_states = residual + feed_forward_hidden_states
+        return hidden_states
+
+
+class JAISModel(nn.Module):
+
+    def __init__(
+        self,
+        config: JAISConfig,
+        linear_method: Optional[LinearMethodBase] = None,
+    ):
+        super().__init__()
+        self.config = config
+        assert not config.add_cross_attention
+        assert not config.scale_attn_by_inverse_layer_idx
+        assert not config.reorder_and_upcast_attn
+        self.embed_dim = config.hidden_size
+        self.wte = VocabParallelEmbedding(config.vocab_size, self.embed_dim)
+        self.wpe = nn.Embedding(
+            config.max_position_embeddings, self.embed_dim
+        ) if config.position_embedding_type != "alibi" else None
+        if hasattr(config, 'embeddings_scale'):
+            self.embeddings_scale = config.embeddings_scale
+        else:
+            self.embeddings_scale = config.mup_embeddings_scale
+        self.h = nn.ModuleList([
+            JAISBlock(config, linear_method)
+            for _ in range(config.num_hidden_layers)
+        ])
+        self.ln_f = nn.LayerNorm(self.embed_dim, eps=config.layer_norm_epsilon)
+
+    def forward(
+        self,
+        input_ids: torch.Tensor,
+        position_ids: torch.Tensor,
+        kv_caches: List[KVCache],
+        input_metadata: InputMetadata,
+    ) -> torch.Tensor:
+        inputs_embeds = self.wte(input_ids)
+        if self.wpe is not None:
+            position_embeds = self.wpe(position_ids)
+            hidden_states = inputs_embeds + position_embeds
+        else:
+            hidden_states = inputs_embeds
+        hidden_states *= torch.tensor(float(self.embeddings_scale),
+                                      dtype=hidden_states.dtype)
+
+        for i in range(len(self.h)):
+            layer = self.h[i]
+            hidden_states = layer(hidden_states, kv_caches[i], input_metadata)
+
+        hidden_states = self.ln_f(hidden_states)
+        return hidden_states
+
+
+class JAISLMHeadModel(nn.Module):
+
+    def __init__(
+        self,
+        config: JAISConfig,
+        linear_method: Optional[LinearMethodBase] = None,
+    ):
+        super().__init__()
+        self.config = config
+        self.linear_method = linear_method
+        self.transformer = JAISModel(config, linear_method)
+        self.lm_head_weight = self.transformer.wte.weight
+        if hasattr(config, 'width_scale'):
+            self.output_logits_scale = config.width_scale
+        else:
+            self.output_logits_scale = config.mup_output_alpha * config.mup_width_scale
+        self.sampler = JAISSampler(config.vocab_size)
```

Comment:
Could you elaborate that why use custom `JAISSampler`? Any differences?

Reply:
We use this as we scale the output logits from lm_head in our base code.

HF code [line ref](https://huggingface.co/core42/jais-30b-chat-v3/blob/main/modeling_jais.py#L1217:~:text=lm_logits%20*%3D%20torch.tensor(float(self.output_logits_scale)%2C%20dtype%3Dlm_logits.dtype%2C%20device%3Dlm_logits.device))
vLLM code [line ref](https://github.com/grandiose-pizza/vllm-jais/blob/65669fc3df45b104525c3ccb8e4a7c681e2508bc/vllm/model_executor/models/jais.py#L96)

I wanted to do this without disturbing the base sample class for conflicts and hence inherited and rewrote the forward func. Hope this clarifies.

Reply:
Thanks for clarification! It's indeed not flexible. I will look into this.

Reply:
Thanks.